libclc/ptx-nvidiacl/libspirv/images/image.cl

```diff
@@ -973,3 +975,1511 @@ _CLC_DEF int3 _Z22__spirv_ImageQuerySizeIDv3_i14ocl_image3d_roET_T0_(
+  return __nvvm_sust_1d_i16_clamp_s(imageHandle, coord, as_short(a));
+}
+void __nvvm_sust_2d_f16_clamp_s(unsigned long imageHandle, int x, int y, half a) {
+  return __nvvm_sust_2d_i16_clamp_s(imageHandle, x, y, as_short(a));
+}
+void __nvvm_sust_3d_f16_clamp_s(unsigned long imageHandle, int x, int y, int z, half a) {
+  return __nvvm_sust_3d_i16_clamp_s(imageHandle, x, y, z, as_short(a));
+}
+void __nvvm_sust_1d_v2f16_clamp_s(unsigned long imageHandle, int coord, half a, half b) {
+  return __nvvm_sust_1d_v2i16_clamp_s(imageHandle, coord, as_short(a), as_short(b));
+}
+void __nvvm_sust_2d_v2f16_clamp_s(unsigned long imageHandle, int x, int y, half a, half b) {
+  return __nvvm_sust_2d_v2i16_clamp_s(imageHandle, x, y, as_short(a), as_short(b));
+}
+void __nvvm_sust_3d_v2f16_clamp_s(unsigned long imageHandle, int x, int y, int z, half a, half b) {
+  return __nvvm_sust_3d_v2i16_clamp_s(imageHandle, x, y, z, as_short(a), as_short(b));
+}
+void __nvvm_sust_1d_v4f16_clamp_s(unsigned long imageHandle, int coord, half a, half b, half c, half d) {
+  return __nvvm_sust_1d_v4i16_clamp_s(imageHandle, coord, as_short(a), as_short(b), as_short(c), as_short(d));
+}
+void __nvvm_sust_2d_v4f16_clamp_s(unsigned long imageHandle, int x, int y, half a, half b, half c, half d) {
+  return __nvvm_sust_2d_v4i16_clamp_s(imageHandle, x, y, as_short(a), as_short(b), as_short(c), as_short(d));
+}
+void __nvvm_sust_3d_v4f16_clamp_s(unsigned long imageHandle, int x, int y, int z, half a, half b, half c, half d) {
+  return __nvvm_sust_3d_v4i16_clamp_s(imageHandle, x, y, z, as_short(a), as_short(b), as_short(c), as_short(d));
+}
+
+#define _CLC_DEFINE_IMAGE_BINDLESS_READ_BUILTIN(                               \
+    elem_t, dimension, elem_t_mangled, vec_size, coord_mangled, coord_input,   \
+    coord_parameter)                                                           \
+  _CLC_DEF                                                                     \
+  elem_t _Z17__spirv_ImageReadI##elem_t_mangled##m##coord_mangled##ET_T0_T1_(  \
+      unsigned long imageHandle, coord_input) {                                \
+    return __nvvm_suld_##dimension##d_##vec_size##_clamp_s(imageHandle,        \
+                                                           coord_parameter);   \
+  }
+
+#define _CLC_DEFINE_IMAGE_BINDLESS_WRITE_BUILTIN(                              \
+    elem_t, dimension, write_mangled, elem_t_mangled, vec_size, coord_input,   \
+    coord_parameter, data_input)                                               \
+  _CLC_DEF void                                                                \
+      _Z18__spirv_ImageWriteIm##write_mangled##elem_t_mangled##EvT_T0_T1_(     \
+          unsigned long imageHandle, coord_input, elem_t c) {                  \
+    __nvvm_sust_##dimension##d_##vec_size##_clamp_s(                           \
+        imageHandle, coord_parameter, data_input);                             \
+  }
+
+#define COMMA ,
```

Comment:
I think with a bit of reordering and using variadic macros (`...` argument and `__VA_ARGS__`) you can remove the need for this `COMMA` macro.

I.e. for `_CLC_DEFINE_IMAGE_BINDLESS_READ_BUILTIN` you should be able to do something like:
```c++
#define _CLC_DEFINE_IMAGE_BINDLESS_READ_BUILTIN(                               \
    elem_t, dimension, elem_t_mangled, vec_size, coord_mangled, coord_input,   \
    ...)                                                           \
  _CLC_DEF                                                                     \
  elem_t _Z17__spirv_ImageReadI##elem_t_mangled##m##coord_mangled##ET_T0_T1_(  \
      unsigned long imageHandle, coord_input) {                                \
    return __nvvm_suld_##dimension##d_##vec_size##_clamp_s(imageHandle,        \
                                                           __VA_ARGS__);   \
  }
```

while for `_CLC_DEFINE_IMAGE_BINDLESS_WRITE_BUILTIN` you could do something similar, but you will need to reorder the arguments so the variadic one is last. Likewise you would have to go through the uses of it and make the appropriate reordering.

Reply:
Noted. We will change to use `__VA_ARGS__`, although for the mipmap macros this might require splitting our current single macro into multiple, at least one for each dimension

Reply:
I've updated the non-mipmap macros to take `__VA_ARGS__` instead of `COMMA` separated arguments.

For the mipmap macros, however, it is not possible as there are multiple `COMMA` separated parameters that feed into different function calls within the macro. We could triplicate all the mipmap macros (per each dimension), but then we would have a very large list of parameters to these macros.

I've come up with the following compromise, which uses a comma in a macro, however I think the end result is much more readable.

```c++
#define COORD_INPUT_1D float x
#define COORD_INPUT_2D float x, float y
#define COORD_INPUT_3D float x, float y, float z

#define COORD_PARAMS_1D x
#define COORD_PARAMS_2D x, y
#define COORD_PARAMS_3D x, y, z

#define GRAD_INPUT_1D float dX, float dY
#define GRAD_INPUT_2D float dXx, float dXy, float dYx, float dYy
#define GRAD_INPUT_3D float dXx, float dXy, float dXz, float dYx, float dYy, float dYz

_CLC_DEFINE_MIPMAP_BINDLESS_THUNK_READS_BUILTIN(float, 1, f32, v4f32, COORD_INPUT_1D, COORD_PARAMS_1D, GRAD_INPUT_1D, dX, dY)
_CLC_DEFINE_MIPMAP_BINDLESS_THUNK_READS_BUILTIN(float, 2, f32, v4f32, COORD_INPUT_2D, COORD_PARAMS_2D, GRAD_INPUT_2D, dXx, dXy, dYx, dYy)
_CLC_DEFINE_MIPMAP_BINDLESS_THUNK_READS_BUILTIN(float, 3, f32, v4f32, COORD_INPUT_3D, COORD_PARAMS_3D, GRAD_INPUT_3D, dXx, dXy, dXz, dYx, dYy, dYz)
```

The last parameters are left as `__VA_ARGS__`.

Let me know what your thoughts are on this. If necessary, we can split the mipmap macros into three (per dimension), but like I said the readability would be much worse in my opinion.

Reply:
I think it looks good, thanks!